flamedisx/lxe_blocks/reconstruct_signals.py

```diff
@@ -0,0 +1,252 @@
+                                    data_tensor=data_tensor,
+                                    bonus_arg=s_raw,
+                                    ptensor=ptensor)
+        # add offset to std to avoid NaNs from norm.pdf if std = 0
+        smear = tf.clip_by_value(relative_smear,
+                                 clip_value_min=1e-15,
+                                 clip_value_max=tf.float32.max)
+
+        result = tfp.distributions.Normal(
+            loc=mu, scale=smear).prob(s_observed)
+
+        # Add detection/selection efficiency
+        result *= self.gimme(self.signal_name + '_acceptance',
+                             data_tensor=data_tensor, ptensor=ptensor)[:, o, o]
+
+        return result
+
+    def check_data(self):
+        if not self.check_acceptances:
+            return
+        s_acc = self.gimme_numpy(self.signal_name + '_acceptance')
+        if np.any(s_acc <= 0):
+            raise ValueError(f"Found event with non-positive {self.signal_name} "
+                             f"acceptance: did you apply and configure "
+                             "your cuts correctly?")
+
+
+@export
+class ReconstructS1(ReconstructSignals):
+
+    raw_signal_name = 's1_raw'
+    signal_name = 's1'
+
+    dimensions = ('s1_raw', 's1')
+    special_model_functions = (
+        'reconstruction_bias_s1_simulate',
+        'reconstruction_smear_s1_simulate',
+        'reconstruction_bias_s1_annotate',
+        'reconstruction_smear_s1_annotate',)
+    model_functions = (
+        's1_acceptance',
+        ) + special_model_functions
+
+    max_dim_size = {'s1_raw': 120}
+    s1_smear_load = 0.01
+
+    def s1_acceptance(self, s1, s1_min=2, s1_max=70):
+        return tf.where((s1 < s1_min) | (s1 > s1_max),
+                        tf.zeros_like(s1, dtype=fd.float_type()),
+                        tf.ones_like(s1, dtype=fd.float_type()))
+
+    # Getting from s1_raw -> s1
+    def reconstruction_bias_s1_simulate(self, s1_raw):
+        """ Dummy method for pax s1 reconstruction bias mean. Overwrite
+        it in source specific class. See x1t_sr1.py for example.
+        s1_raw is a float64, output is fd.float_type() which is a float32
+        """
+        return tf.ones_like(s1_raw, dtype=fd.float_type())
+
+    def reconstruction_smear_s1_simulate(self, s1_raw):
+        """ Dummy method for pax s1 reconstruction bias spread. Overwrite
+        it in source specific class. See x1t_sr1.py for example.
+
+        Not quite a dirac delta, which will make this block an identity matrix
+        but computationally intractible or at least not trivially. Need to smear
+        this dirac delta out by a small loading term of 0.001. A larger s1_raw
+        max_dim_size would need a smaller loading term.
+        """
+        return tf.zeros_like(s1_raw, dtype=fd.float_type())+self.s1_smear_load
+
+    # Getting from s1 -> s1_raw
+    def reconstruction_bias_s1_annotate(self, s1_raw):
```

Comment:
Similar to comment to that above... Maybe clarify it in the code, also?

Reply:
Yes, the code defines two pairs of model functions. If we write docs for this it would be good to emphasize this.